### ISSUE TITLE
RFC: session probes

### ISSUE THREAD

**Issue by viktormalik** [bpftrace/bpftrace#3525]
### Is your feature request related to a problem? Please describe.

Kernel 6.10 has introduced so-called "kprobe session probes" ([patch series](https://lore.kernel.org/bpf/20240430112830.1184228-1-jolsa@kernel.org/)) which are an extension of kprobe-multi that allow a single BPF program attached to both entry and exit of the function. The `bpf_session_is_return` kfunc can be used to determine if the current attachment is in the entry or the exit probe. In addition, the entry and the exit probes can share data via a "cookie" (available via the `bpf_session_cookie` kfunc).

Once we have kfuncs support (#3523), we should be able to use session probes. This RFC is to discuss the necessary syntax decisions.

### Describe the solution you'd like

There are two solutions I currently have in mind:

1. Introduce a new probe type

    This should allow this simple program to compute average runtime of each VFS function:
    ```
    ksession:vfs_*
    {
        if (is_return())
            cookie = nsecs;
        else
            @runtime[func] = avg(nsecs - cookie);
    }
    ```

2. Detect if user uses `kprobe:X` and `kretprobe:X` and transparently merge his probes into one session probe. This is purely for peformance benefits as session probes are faster, especially when using wildcard attachment.

Overall, I can see benefits in both approaches. Session probes are mostly about performance and the second one uses it transparently without user even noticing, however, it fails to leverage all the benefits (sharing code between the probes, usage of cookie) so the first approach can be even better but requires the user to explicitly use it.

**Comment by ajor** [comment]
I'll put my vote behind the second option (transparently use session probes where applicable). It'll benefit all existing scripts and we won't have to tell users to forget the old kprobe/kretprobe patterns and to use this new probe type on newer kernels.

Code sharing we'll soon support through user defined functions anyway. I suppose the cookie will perform better, but we can already achieve similar things with maps, plus the cookie is more limited (u64 only).

**Comment by jordalgo** [comment]
I'm also for 2nd option for now (for the reasons @ajor mentioned). I think we can punt on the decision to add a new probe type later when there is more of an explicit need and it's available on more kernels/distros.

**Comment by batu3H** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by viktormalik** [comment]
Option 2 is implemented in #3772. I'm still considering option 1 as it may have other performance benefits which are not available (or hard to achieve) via option 2. I'll do some measurements and come back. Until then, I'd keep this issue open.

**Comment by amscanne** [comment]
Related to #4114 as the probe matching can be more effective with early expansion.

**Comment by t-8ch** [comment]
> I suppose the cookie will perform better, but we can already achieve similar things with maps, plus the cookie is more limited (u64 only).

With the current infrastructure it doesn't seem to be possible to match entry/exit for a function that can be executed concurrently *within a single task*. For example because it is executed by a IRQ handler while the regular task which is interrupted by said IRQ handler is currently executing the same function function. In this case the normal pattern of `@map[tid]` corrupts the map, as the 'tid' of the IRQ handler is the same as the task it interrupted.

Using a unique session cookie as a map key instead should avoid this.

**Comment by jordalgo** [comment]
@t-8ch Good point. We'd probably have to make some changes to allow users to explicitly opt-in to session probes before exposing something to write/read the cookie. I'm thinking:
```
ksession:do_nanosleep {
    if (!is_return) {
      set_cookie(nsecs);
    } else {
      printf("latency: %d ns\n", nsecs - get_cookie());
    }
  }
```
This shouldn't be hard to do and we can do feature checks to make sure the session probe is available.

**Comment by t-8ch** [comment]
Using `nsecs` as cookie is sufficient to compute the latency.
However for more complex schemes it would be good to have a guaranteed unique cookie, as `nsec` is not collision-proof.
Either by preseeding the cookie with a value that is guaranteed to be unique during the lifetime of the session or by giving the user access to an atomic counter, maybe in the form of a special map.

Out of curiousity:
What would prevent the usage of the cookie from the normal `kprobe`/`kretprobe` probes which are implicitly converted to sessions? The opt-in would be the fact that the probe uses the cookie variable.

**Comment by jordalgo** [comment]
> What would prevent the usage of the cookie from the normal kprobe/kretprobe probes which are implicitly converted to sessions? The opt-in would be the fact that the probe uses the cookie variable.

That's true. I think the weirdness I wanted to avoid with an explicit session probe was in the case of multi-attachpoints - if a probe uses the new get/set cookie builtin it has to be valid in all contexts when we do probe expansion. cc @viktormalik 

> However for more complex schemes it would be good to have a guaranteed unique cookie, as nsec is not collision-proof.
Either by preseeding the cookie with a value that is guaranteed to be unique during the lifetime of the session or by giving the user access to an atomic counter, maybe in the form of a special map.

Yeah, I guess a regular per-cpu map would probably just fail to get the lock if it's already held on the same cpu. IIUC.

**Comment by viktormalik** [comment]
> > What would prevent the usage of the cookie from the normal kprobe/kretprobe probes which are implicitly converted to sessions? The opt-in would be the fact that the probe uses the cookie variable.
> 
> That's true. I think the weirdness I wanted to avoid with an explicit session probe was in the case of multi-attachpoints - if a probe uses the new get/set cookie builtin it has to be valid in all contexts when we do probe expansion. cc [@viktormalik](https://github.com/viktormalik)

Yeah, I guess that's my concern, too. When doing automatic transformation of kprobe/kretprobe pairs to session probes, we'd have to make sure that cookie is only used in probes that will be auto-converted. With explicit session probes, we could omit that check.
